### ISSUE TITLE
Error: Mask in MaskedCrossEntropyLoss is all zeros

### ISSUE THREAD

**Issue by arbab97** [wvangansbeke/Unsupervised-Classification#46]
I get error in third (self-labeling) step. The error occurs at these lines:

 def forward(self, input, target, mask, weight, reduction='mean'):
        if not (mask != 0).any():
            raise ValueError('Mask in MaskedCrossEntropyLoss is all zeros.')

Any suggestions?

**Comment by arbab97** [comment]
Update: 
I had to comment the lines from losses.py. Now it is executing but I am not sure if this is wrong ....

class MaskedCrossEntropyLoss(nn.Module):
    def __init__(self):
        super(MaskedCrossEntropyLoss, self).__init__()
        
    def forward(self, input, target, mask, weight, reduction='mean'):
        #!!!!!!!!!!!! ALL COMMENTED
        # if not (mask != 0).any():
        #     raise ValueError('Mask in MaskedCrossEntropyLoss is all zeros.')
        # target = torch.masked_select(target, mask)
        # b, c = input.size()
        # n = target.size(0)
        # input = torch.masked_select(input, mask.view(b, 1)).view(n, c)

        return F.cross_entropy(input, target, weight = weight, reduction = reduction)

**Comment by wvangansbeke** [comment]
Hi @arbab97,

We only backpropagate the loss for confident samples during the self-labeling step (see [this line](https://github.com/wvangansbeke/Unsupervised-Classification/blob/master/losses/losses.py#L43)). 
If you want to see improvement, you should keep the mask.

The error tells you that it can not find any confident samples. So, try increasing the batch size or reducing the confidence threshold.
On which dataset are you working and what batch size are you using?
If you use our settings, it should not be a problem.  

Hope this helps.

**Comment by arbab97** [comment]
Thank you very much. Right now due to limitation of my GPU , I was using batch size of 1 and there's a custom datasets that I am using. 
"The error tells you that it can not find any confident samples" <- I think this is the reason then. 
Thank you,

**Comment by adrielkuek** [comment]
Hi, sorry just to jump on this bandwagon regarding the MaskedCrossEntropyLoss. So essentially if commented out the mask portion would render step 3 self-label training as redundant?

Thanks!

> Hi @arbab97,
> 
> We only backpropagate the loss for confident samples during the self-labeling step (see [this line](https://github.com/wvangansbeke/Unsupervised-Classification/blob/master/losses/losses.py#L43)).
> If you want to see improvement, you should keep the mask.
> 
> The error tells you that it can not find any confident samples. So, try increasing the batch size or reducing the confidence threshold.
> On which dataset are you working and what batch size are you using?
> If you use our settings, it should not be a problem.
> 
> Hope this helps.

**Comment by wvangansbeke** [comment]
Hi @adrielkuek,

Yes. Not using a mask can even worsen the results, since you might enforce mistakes. We observed that high confident samples are more likely to be correctly classified. These can be used as prototypes. You need a confidence mask, (ideally at 95+% confidence) to readjust the classification boundary, conditioned on the prototypes of a particular class.

**Comment by adrielkuek** [comment]
> Hi @adrielkuek,
> 
> 
> 
> Yes. Not using a mask can even worsen the results, since you might enforce mistakes. We observed that high confident samples are more likely to be correctly classified. These can be used as prototypes. You need a confidence mask, (ideally at 95+% confidence) to readjust the classification boundary, conditioned on the prototypes of a particular class. 

Thanks! Appreciate the insights on the prototypical behavior. I'll experiment varying the confidence mask to see which provides the best classification boundary. 👍🏻

**Comment by MotiBaadror** [comment]
How to proceed when I do not have labels for the images, so I am keeping all the images in one folder, I kept the target variable as 0 for dataloader to work. Now in step 3 I am getting the above error. Am I missing something or how can I make it work for the images where labels are not available.
